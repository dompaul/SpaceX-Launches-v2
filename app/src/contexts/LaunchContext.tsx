import React from "react";
import { LaunchOption } from "../models/LaunchModel";
import { getLaunchesAPI } from "../api/getLaunches";
import { getYears } from "../utils/getYears";

export interface LaunchContextProps {
  listLaunches: Function;
  items: LaunchOption[];
  loaded: boolean;
  error: boolean;
  sort: boolean;
  setSort: Function;
  years: number[];
  filter: string;
  setFilter: Function;
}

export const launchContextDefaults: LaunchContextProps = {
  listLaunches: Function,
  items: [],
  loaded: false,
  error: false,
  sort: false,
  setSort: Function,
  years: [],
  filter: "",
  setFilter: Function,
};

export const LaunchContext = React.createContext<LaunchContextProps>(
  launchContextDefaults
);
export const useLaunchContext = (): LaunchContextProps =>
  React.useContext(LaunchContext);

export const LaunchProvider: React.FC = ({ children }) => {
  const [items, setItems] = React.useState<LaunchOption[]>([]);
  const [loaded, setLoading] = React.useState<boolean>(false);
  const [error, setError] = React.useState<boolean>(false);
  const [sort, setSort] = React.useState<boolean>(false);
  const [years, setYears] = React.useState<number[]>([]);
  const [filter, setFilter] = React.useState<string>("");

  return (
    <LaunchContext.Provider
      value={{
        listLaunches: React.useCallback(async () => {
          try {
            setLoading(false);
            setError(false);
            setFilter("");
            const response = await getLaunchesAPI();
            setItems(response);
            setYears(getYears(response));
            setLoading(true);
          } catch {
            setError(true);
          }
        }, []),
        items,
        loaded,
        error,
        sort,
        setSort,
        years,
        filter,
        setFilter,
      }}
    >
      {children}
    </LaunchContext.Provider>
  );
};
