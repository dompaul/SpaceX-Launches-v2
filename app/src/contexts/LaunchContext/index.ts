export { useLaunchContext } from "./LaunchContext";
export { LaunchContextMock } from "./LaunchContextMock";
