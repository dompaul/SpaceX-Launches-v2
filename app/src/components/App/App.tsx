import React from "react";
import { Header } from "../../components/Header";
import { Body } from "../../components/Body";
import { useLaunchContext } from "../../contexts/LaunchContext";

export const App: React.FC = () => {
  const { listLaunches } = useLaunchContext();

  React.useEffect(() => {
    listLaunches();
  }, [listLaunches]);

  return (
    <div className="app">
      <Header />
      <Body />
    </div>
  );
};
