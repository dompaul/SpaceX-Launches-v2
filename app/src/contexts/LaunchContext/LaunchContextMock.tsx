import React from "react";
import { LaunchOption } from "../../models/LaunchModel";
import { LaunchContext, launchContextDefaults } from "./LaunchContext";

export interface LaunchContextMockProps {
  listLaunches?: Function;
  items?: LaunchOption[];
  loaded?: boolean;
  error?: boolean;
  sort?: boolean;
  setSort?: Function;
  years?: number[];
  filter?: string;
  setFilter?: Function;
}

export const LaunchContextMock: React.FC<LaunchContextMockProps> = ({
  children,
  ...overrides
}) => {
  return (
    <LaunchContext.Provider value={{ ...launchContextDefaults, ...overrides }}>
      {children}
    </LaunchContext.Provider>
  );
};
