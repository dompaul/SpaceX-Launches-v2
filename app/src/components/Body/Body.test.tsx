import {
  getByTestId,
  render,
  RenderResult,
  fireEvent,
} from "@testing-library/react";
import { App } from "../App";
import LABEL from "../../constants/Labels";
import { LaunchContextMock } from "../../contexts/LaunchContext";

const testIds = {
  sortButton: "sort-button-test",
  filterButton: "filter-button-test",
  launchList: "launch-list-test",
  loading: "loading-state-test",
  error: "error-state-test",
};

describe("Body.tsx", () => {
  let wrapper: RenderResult;
  const mockList = jest.fn();
  const mockSort = jest.fn();

  describe("default", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchContextMock listLaunches={mockList}>
          <App />
        </LaunchContextMock>
      );
    });

    it("renders with default snapshot", () => {
      expect(wrapper.container).toMatchSnapshot();
    });

    it("renders filter button component", () => {
      const filterButton = getByTestId(wrapper.container, testIds.filterButton);
      expect(filterButton).toBeInTheDocument();
    });

    it("renders sort button component", () => {
      const sortButton = getByTestId(wrapper.container, testIds.sortButton);
      expect(sortButton).toBeInTheDocument();
    });
  });

  describe("on load", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchContextMock listLaunches={mockList}>
          <App />
        </LaunchContextMock>
      );
    });

    it("renders loading state", () => {
      const loadingMessage = getByTestId(wrapper.container, testIds.loading);
      expect(loadingMessage).toHaveTextContent(LABEL.LOADING);
    });
  });

  describe("on error", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchContextMock error={true} listLaunches={mockList}>
          <App />
        </LaunchContextMock>
      );
    });

    it("renders error state", () => {
      const errorMessage = getByTestId(wrapper.container, testIds.error);
      expect(errorMessage).toHaveTextContent(LABEL.ERROR);
    });
  });

  describe("on click", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchContextMock
          setSort={mockSort}
          sort={false}
          loaded={true}
          listLaunches={mockList}
        >
          <App />
        </LaunchContextMock>
      );
    });

    it("sort button", () => {
      fireEvent.click(getByTestId(wrapper.container, testIds.sortButton));
      expect(mockSort).toHaveBeenCalledTimes(1);
    });
  });

  describe("sort: descending order", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchContextMock
          setSort={mockSort}
          sort={false}
          loaded={true}
          listLaunches={mockList}
        >
          <App />
        </LaunchContextMock>
      );
    });

    it("sorts list descending", () => {
      expect(
        getByTestId(wrapper.container, testIds.sortButton)
      ).toHaveTextContent(LABEL.DESC);
    });
  });

  describe("sort: ascending order", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchContextMock
          setSort={mockSort}
          sort={true}
          loaded={true}
          listLaunches={mockList}
        >
          <App />
        </LaunchContextMock>
      );
    });

    it("sorts list ascending", () => {
      expect(
        getByTestId(wrapper.container, testIds.sortButton)
      ).toHaveTextContent(LABEL.ASC);
    });
  });
});
