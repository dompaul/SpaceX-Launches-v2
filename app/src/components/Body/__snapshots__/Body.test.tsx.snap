// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body.tsx default renders with default snapshot 1`] = `
<div>
  <div
    class="app"
  >
    <header
      class="app__header"
    >
      <div
        class="app__logo"
      >
        <img
          alt="SpaceX"
          class="app__logo-image"
          src="spacex-logo.png"
        />
        <span
          class="app__logo-txt"
        >
          Launches
        </span>
      </div>
      <button
        class="button button--reload"
        data-testid="reload-button-test"
      >
        Reload Data
      </button>
    </header>
    <div
      class="app__body"
    >
      <div
        class="app__image-container"
      >
        <img
          alt="Launch Home"
          class="app__main-image"
          src="launch-home.png"
          srcset="launch-home.png 1x, launch-home@2x.png 2x, launch-home@3x.png 3x"
        />
      </div>
      <div
        class="app__launches"
      >
        <div
          class="app__filters"
        >
          <select
            class="select"
            data-testid="filter-button-test"
            name="Filter By Year"
          >
            <option
              value=""
            >
              Filter By Year
            </option>
          </select>
          <button
            class="button button--sort"
            data-testid="sort-button-test"
          >
            Sort Descending
          </button>
        </div>
        <span
          data-testid="loading-state-test"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
